kinesis_consumer.cpp

```diff
@@ -0,0 +1,266 @@
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/time.h>
+
+using namespace Aws::Auth;
+using namespace Aws::Client;
+using namespace Aws::Kinesis;
+using namespace Aws::Kinesis::Model;
+using namespace Aws::Utils;
+
+struct kinesis_consumer
+{
+	KinesisClient *client;
+	concurrent_queue<GetRecordsOutcome*> *queue;
+	std::atomic<bool> keep_running;
+	std::thread *thread;
+	Aws::String shard_iter;
+};
+
+// Override the default logging system to print to stderr
+class AltLogSystem : public Logging::FormattedLogSystem
+{
+  public:
+	AltLogSystem(Logging::LogLevel logLevel) : FormattedLogSystem(logLevel) {}
+	virtual ~AltLogSystem() {}
+
+  protected:
+	virtual void ProcessFormattedStatement(Aws::String&& statement)
+	{
+		fprintf(stderr, "%s", statement.c_str());
+	}
+};
+
+static double
+get_time()
+{
+    struct timeval tv;
+    gettimeofday(&tv, 0);
+    return tv.tv_sec + (tv.tv_usec / 1000000.0);
+}
+
+static void consume_thread(kinesis_consumer *kc);
+
+// create the state for the consumer.
+// note - this is a blocking call wrt obtaining the shard iterator.
+// on error this returns NULL.
+kinesis_consumer*
+kinesis_consumer_create()
+{
+	Aws::Utils::Logging::InitializeAWSLogging(Aws::MakeShared<AltLogSystem>("logging", Aws::Utils::Logging::LogLevel::Error));
+
+	ClientConfiguration config;
+	config.region = Aws::Region::US_WEST_2;
+	config.verifySSL = false;
+	
+	config.retryStrategy = Aws::MakeShared<DefaultRetryStrategy>("kinesis_consumer", 5, 25);
+	auto *kc = new KinesisClient(config);
+
+	GetShardIteratorRequest request;
+
+	request.SetStreamName("test");
+	request.SetShardId("shardId-000000000000");
+	request.SetShardIteratorType(ShardIteratorType::TRIM_HORIZON);
+
+	auto outcome = kc->GetShardIterator(request);
+
+	if (!outcome.IsSuccess())
+	{
+		delete kc;
+		return NULL;
+	}
+
+	Aws::String shard_iter = outcome.GetResult().GetShardIterator();
+
+	auto *cq = new concurrent_queue<GetRecordsOutcome*>(100);
+	return new kinesis_consumer{kc, cq, {true}, NULL, shard_iter};
+}
+
+// create and start the consumer thread
+void
+kinesis_consumer_start(kinesis_consumer *kc)
+{
+	kc->thread = new std::thread(consume_thread, kc);
+}
+
+// stop, join, and delete the consumer thread
+void
+kinesis_consumer_stop(kinesis_consumer *kc)
+{
+	kc->keep_running = false;
+	kc->thread->join();
+
+	delete kc->thread;
+}
+
+// clean up consumer resources
+void
+kinesis_consumer_destroy(kinesis_consumer *kc)
+{
+	if (kc->keep_running)
+		kinesis_consumer_stop(kc);
+
+	delete kc->queue;
+	delete kc->client;
+	delete kc;
+}
+
+// consumer thread continuously loops requesting records until the shard
+// is finished or it is signalled to stop (via kc->keep_running).
+static void
+consume_thread(kinesis_consumer *kc)
+{
+	KinesisClient &client = *kc->client;
+
+	double last_request_time = 0.0;
+
+	GetRecordsRequest req;
+
+	while (kc->keep_running)
+	{
+		req.SetShardIterator(kc->shard_iter);
+		req.SetLimit(1000);
```

Comment:
This limit should probably come from a member of the `kinesis_consumer`, even if it's hard-coded as `1000` for now when the consumer is created.